File: tests/unit/import.test.ts
// The import (F16): the envelope guard, the duplicate rule, and the one thing
// that must never cross the boundary — an id.
//
// Every bundle here is written out by hand, because that is what a bundle IS.
// This app has no export: the file always comes from somewhere else, and a test
// that fed the importer its own output would be proving the wrong thing.

import { existsSync, mkdtempSync, readFileSync, readdirSync, rmSync, writeFileSync } from 'node:fs'
import { tmpdir } from 'node:os'
import { join } from 'node:path'
import { afterEach, beforeEach, describe, expect, it } from 'vitest'
import { BundleError } from '../../src/shared/bundle.js'
import { importBundle } from '../../src/main/storage/import.js'
import { loadVehicle, vehicleFiles } from '../../src/main/storage/repository.js'

let home = ''
let previous: string | undefined

beforeEach(() => {
  home = mkdtempSync(join(tmpdir(), 'tritium-import-'))
  previous = process.env['XDG_DATA_HOME']
  process.env['XDG_DATA_HOME'] = home
})

afterEach(() => {
  if (previous === undefined) delete process.env['XDG_DATA_HOME']
  else process.env['XDG_DATA_HOME'] = previous
  rmSync(home, { recursive: true, force: true })
})

function bundleFile(text: string): string {
  const file = join(home, 'bundle.toml')
  writeFileSync(file, text)
  return file
}

const ONE_FILL = `
format = "tritium-export"
format_version = 1
exported = 2026-08-19

[[vehicle]]
slug = "sportage"
name = "SPORTAGE 1.6 T-GDI"
tank_capacity_l = 54.0

[[vehicle.fuel]]
date = 2026-08-16
odometer_km = 19764
litres = 29.990
price_per_litre = 73.380
full_tank = true
fuel_type = "Kurşunsuz 95"
`

describe('the envelope', () => {
  it('refuses a file that is not a bundle', () => {
    const file = bundleFile('schema_version = 1\n\n[[entry]]\nid = "f-0001"\n')
    expect(() => importBundle(file)).toThrow(BundleError)
    // A record file is not a bundle even though it parses perfectly.
    expect(existsSync(join(home, 'tritium', 'vehicles'))).toBe(false)
  })

  it('refuses TOML that will not parse, and writes nothing', () => {
    const file = bundleFile('format = = =\n')
    expect(() => importBundle(file)).toThrow(BundleError)
    expect(existsSync(join(home, 'tritium', 'vehicles'))).toBe(false)
  })

  /*
   * The guard that does not exist for record files, and must exist here.
   *
   * The storage layer reads schema_version and then writes back the current
   * constant whatever it read. Inside one app that is an upgrade. Across two
   * apps it is a silent downgrade — the phone's newer format relabelled as this
   * one's older format, with the difference lost.
   */
  it('refuses a bundle newer than it understands rather than downgrading it', () => {
    const file = bundleFile(ONE_FILL.replace('format_version = 1', 'format_version = 2'))
    expect(() => importBundle(file)).toThrow(BundleError)
    expect(existsSync(join(home, 'tritium', 'vehicles'))).toBe(false)
    // And the bundle itself is untouched — we refuse files, we do not edit them.
    expect(readFileSync(file, 'utf8')).toContain('format_version = 2')
  })
})

describe('importing', () => {
  it('creates a vehicle that is not here yet, and its fill-up', () => {
    const result = importBundle(bundleFile(ONE_FILL))

    expect(result.vehicles).toHaveLength(1)
    expect(result.vehicles[0]?.vehicleCreated).toBe(true)
    expect(result.vehicles[0]?.added.fuel).toBe(1)

    const bundle = loadVehicle('sportage')
    expect(bundle.vehicle?.vehicle.name).toBe('SPORTAGE 1.6 T-GDI')
    expect(bundle.fuel.entries).toHaveLength(1)
    // Scaled on the way in by the record file's own reader, not by this module.
    expect(bundle.fuel.entries[0]?.litres).toBe(29990)
    expect(bundle.fuel.entries[0]?.odometer_km).toBe(19764)
  })

  it('gives the entry an id of its own, and the bundle carries none', () => {
    importBundle(bundleFile(ONE_FILL))
    expect(loadVehicle('sportage').fuel.entries[0]?.id).toBe('f-0001')
  })

  /*
   * The whole point of the milestone. A monthly sync overlaps the previous one,
   * and importing the same file twice must be a no-op rather than a doubling.
   */
  it('imports the same bundle twice and changes nothing the second time', () => {
    const file = bundleFile(ONE_FILL)
    importBundle(file)
    const after = readFileSync(vehicleFiles('sportage').fuel, 'utf8')

    const second = importBundle(file)
    expect(second.vehicles[0]?.added.fuel).toBe(0)
    expect(second.vehicles[0]?.skipped.fuel).toBe(1)
    expect(readFileSync(vehicleFiles('sportage').fuel, 'utf8')).toBe(after)
  })

  it('adds a new fill-up beside one it already has', () => {
    importBundle(bundleFile(ONE_FILL))

    const more = `${ONE_FILL}
[[vehicle.fuel]]
date = 2026-09-02
odometer_km = 20310
litres = 31.500
price_per_litre = 74.100
full_tank = true
fuel_type = "Kurşunsuz 95"
`
    const result = importBundle(bundleFile(more))
    expect(result.vehicles[0]?.added.fuel).toBe(1)
    expect(result.vehicles[0]?.skipped.fuel).toBe(1)

    const entries = loadVehicle('sportage').fuel.entries
    expect(entries.map((entry) => entry.id)).toEqual(['f-0001', 'f-0002'])
  })

  /*
   * An id in a bundle is ignored rather than honoured. If it were honoured, two
   * devices' `f-0005` would land in one file — and duplicate ids there corrupt
   * it silently, because entryRest is keyed by id and findIndex only ever
   * reaches the first.
   */
  it('ignores an id the bundle tries to bring with it', () => {
    const withId = ONE_FILL.replace(
      '[[vehicle.fuel]]\ndate',
      '[[vehicle.fuel]]\nid = "f-0999"\ndate'
    )
    importBundle(bundleFile(withId))
    expect(loadVehicle('sportage').fuel.entries[0]?.id).toBe('f-0001')
  })

  /*
   * A regression, and it shipped in this milestone's own first draft.
   *
   * A bundle's [[vehicle]] table carries `slug` and the three entry arrays. The
   * vehicle reader carries every key it does not recognise into `rest`, and the
   * writer renders `rest` inline — so the whole fill-up history was landing in
   * vehicle.toml as one array:
   *
   *   fuel = [{ date = 2026-08-16, odometer_km = 19764, litres = 29.99, … }]
   *
   * The tests written first all passed, because every one of them asserted what
   * vehicle.toml SHOULD contain and none asserted what it should not.
   */
  it('does not carry the bundle’s structure into vehicle.toml', () => {
    importBundle(bundleFile(ONE_FILL))
    const written = readFileSync(vehicleFiles('sportage').vehicle, 'utf8')

    expect(written).toContain('SPORTAGE 1.6 T-GDI')
    expect(written).not.toContain('litres')
    expect(written).not.toContain('fuel =')
    expect(written).not.toContain('slug')
  })

  it('leaves an existing vehicle record alone', () => {
    importBundle(bundleFile(ONE_FILL))
    const renamed = ONE_FILL.replace('SPORTAGE 1.6 T-GDI', 'SOMETHING ELSE')
    importBundle(bundleFile(renamed))

    // A bundle adds entries to a vehicle. It does not rewrite the vehicle.
    expect(loadVehicle('sportage').vehicle?.vehicle.name).toBe('SPORTAGE 1.6 T-GDI')
  })
})

describe('the duplicate key, per kind', () => {
  const costs = (category: string, amount: string): string => `
format = "tritium-export"
format_version = 1

[[vehicle]]
slug = "sportage"
name = "SPORTAGE"

[[vehicle.costs]]
date = 2026-01-20
group = "tekrar-eden"
category = "${category}"
title = "Trafik Sigortası"
amount = ${amount}
income = false
`

  it('matches a cost on date, category and amount — it has no odometer', () => {
    importBundle(bundleFile(costs('trafik-sigortasi', '11746.00')))

    const same = importBundle(bundleFile(costs('trafik-sigortasi', '11746.00')))
    expect(same.vehicles[0]?.skipped.costs).toBe(1)

    const other = importBundle(bundleFile(costs('kasko', '11746.00')))
    expect(other.vehicles[0]?.added.costs).toBe(1)
  })

  const service = (part: string): string => `
format = "tritium-export"
format_version = 1

[[vehicle]]
slug = "sportage"
name = "SPORTAGE"

[[vehicle.service]]
date = 2025-05-14
part = "${part}"
odometer_km = 370
amount = 8664.00
`

  /* Tyres AND an oil change on one day at one reading is perfectly ordinary,
   * which is why `part` is in the key and date+odometer alone would be wrong. */
  it('keeps two service records made the same day at the same reading', () => {
    importBundle(bundleFile(service('Michelin Primacy 4')))
    const second = importBundle(bundleFile(service('Motor yağı')))

    expect(second.vehicles[0]?.added.service).toBe(1)
    expect(loadVehicle('sportage').service.entries).toHaveLength(2)
  })
})

describe('backups', () => {
  it('copies what it is about to overwrite, and not what it creates', () => {
    // First import creates the vehicle: there was nothing to preserve.
    importBundle(bundleFile(ONE_FILL))
    expect(existsSync(join(home, 'tritium', 'backups'))).toBe(false)

    const more = `${ONE_FILL}
[[vehicle.fuel]]
date = 2026-09-02
odometer_km = 20310
litres = 31.500
price_per_litre = 74.100
full_tank = true
fuel_type = "Kurşunsuz 95"
`
    importBundle(bundleFile(more))

    const rounds = readdirSync(join(home, 'tritium', 'backups'))
    expect(rounds).toHaveLength(1)

    // The copy is the file as it was BEFORE the second import — one entry.
    const kept = readFileSync(
      join(home, 'tritium', 'backups', rounds[0] as string, 'vehicles', 'sportage', 'fuel.toml'),
      'utf8'
    )
    expect(kept).toContain('f-0001')
    expect(kept).not.toContain('f-0002')
  })
})

File: android/app/src/test/kotlin/io/github/sudomegas/tritium/storage/ImportTest.kt
package io.github.sudomegas.tritium.storage

import java.io.File
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder

/**
 * [VehicleRepository.importBundle], against hand-typed bundle text only —
 * never [Bundle.build]'s own output. F16's own `import.test.ts` never fed
 * the importer its own writer's output either: "a test that fed the
 * importer its own output would be proving the wrong thing." These
 * fixtures are typed the way a maker's editor — or the desktop's own
 * exporter, or another phone's — would produce them, matching F16 §2.2's
 * worked example.
 */
class ImportTest {

    @get:Rule
    val tmp = TemporaryFolder()

    private fun repository(): VehicleRepository = VehicleRepository(TritiumPaths(tmp.root))
    private fun paths(): TritiumPaths = TritiumPaths(tmp.root)

    private val newVehicleBundle = """
        format = "tritium-export"
        format_version = 1
        exported = 2026-08-19
        source = "android"

        [[vehicle]]
        slug = "sportage"
        name = "SPORTAGE 1.6 T-GDI"
        make = "Kia"
        model = "Sportage"
        year = 2025
        engine = "1.6 T-GDI"
        fuel_spec = "Kurşunsuz 95"
        plate = ""
        vin = ""
        tank_capacity_l = 54.0
        purchase_date = 2025-04-25
        purchase_price = 2160000.00
        registration_date = 2025-04-26
        inspection_due = 2027-04-01

        [[vehicle.fuel]]
        date = 2026-08-16
        odometer_km = 19764
        litres = 29.990
        price_per_litre = 73.380
        full_tank = true
        fuel_type = "Kurşunsuz 95"

        [[vehicle.costs]]
        date = 2026-04-11
        group = "tekrar-eden"
        category = "trafik-sigortasi"
        title = "Trafik Sigortası 26/27"
        amount = 11746.00
        income = false
        payment_method = "kredi-karti"
        bank = "Enpara"
        instalment = "Taksit 6"
        note = ""

        [[vehicle.service]]
        date = 2025-05-14
        part = "Michelin Primacy 4 S1 235/50R19 103V XL"
        odometer_km = 370
        amount = 8664.00
        vendor = "https://www.lastikcim.com.tr/"
    """.trimIndent()

    @Test
    fun `importing a bundle creates a new vehicle with fresh ids`() {
        val repository = repository()
        val result = repository.importBundle(newVehicleBundle)

        assertEquals(1, result.vehicles.size)
        val tally = result.vehicles.single()
        assertEquals("sportage", tally.slug)
        assertTrue(tally.vehicleCreated)
        assertEquals(Counts(1, 1, 1), tally.added)
        assertEquals(Counts(0, 0, 0), tally.skipped)

        val bundle = repository.loadVehicle("sportage")
        assertEquals("SPORTAGE 1.6 T-GDI", bundle.vehicle?.vehicle?.name)
        assertEquals("f-0001", bundle.fuel.entries.single().id)
        assertEquals("c-0001", bundle.costs.entries.single().id)
        assertEquals("s-0001", bundle.service.entries.single().id)
    }

    @Test
    fun `the created vehicle_toml never carries an inline fuel, costs, service or slug key`() {
        val repository = repository()
        repository.importBundle(newVehicleBundle)

        val text = paths().vehicleToml("sportage").readText()
        assertFalse(text.contains("fuel ="))
        assertFalse(text.contains("costs ="))
        assertFalse(text.contains("service ="))
        assertFalse(text.contains("slug"))
    }

    @Test
    fun `an existing vehicle's own vehicle_toml is left alone — a bundle only adds entries`() {
        val repository = repository()
        repository.saveVehicleRecord("sportage", VehicleDocument(1, EMPTY_VEHICLE.copy(name = "My Own Name"), emptyMap()))
        val before = paths().vehicleToml("sportage").readText()

        val tally = repository.importBundle(newVehicleBundle).vehicles.single()

        assertFalse(tally.vehicleCreated)
        assertEquals(before, paths().vehicleToml("sportage").readText())
        assertEquals("My Own Name", repository.loadVehicle("sportage").vehicle?.vehicle?.name)
        assertEquals(1, repository.loadVehicle("sportage").fuel.entries.size)
    }

    @Test
    fun `re-importing the same bundle adds nothing the second time — every file byte-identical`() {
        val repository = repository()
        repository.importBundle(newVehicleBundle)
        val fuelBefore = paths().fuelToml("sportage").readText()
        val costsBefore = paths().costsToml("sportage").readText()
        val serviceBefore = paths().serviceToml("sportage").readText()
        val vehicleBefore = paths().vehicleToml("sportage").readText()

        val tally = repository.importBundle(newVehicleBundle).vehicles.single()

        assertFalse(tally.vehicleCreated)
        assertEquals(Counts(0, 0, 0), tally.added)
        assertEquals(Counts(1, 1, 1), tally.skipped)
        assertEquals(fuelBefore, paths().fuelToml("sportage").readText())
        assertEquals(costsBefore, paths().costsToml("sportage").readText())
        assertEquals(serviceBefore, paths().serviceToml("sportage").readText())
        assertEquals(vehicleBefore, paths().vehicleToml("sportage").readText())
    }

    @Test
    fun `an id present in a bundle entry is ignored — the receiving file allocates its own`() {
        val repository = repository()
        val bundle = """
            format = "tritium-export"
            format_version = 1
            exported = 2026-08-19
            source = "android"

            [[vehicle]]
            slug = "spare"
            name = "Spare"

            [[vehicle.fuel]]
            id = "f-9999"
            date = 2026-01-01
            odometer_km = 100
            litres = 10.000
            price_per_litre = 40.000
            full_tank = true
        """.trimIndent()

        repository.importBundle(bundle)
        assertEquals("f-0001", repository.loadVehicle("spare").fuel.entries.single().id)
    }

    @Test
    fun `fuel dedup key is date plus odometer — two fill-ups same day, different reading, both import`() {
        val repository = repository()
        val bundle = """
            format = "tritium-export"
            format_version = 1
            exported = 2026-08-19
            source = "android"

            [[vehicle]]
            slug = "spare"
            name = "Spare"

            [[vehicle.fuel]]
            date = 2026-01-01
            odometer_km = 100
            litres = 10.000
            price_per_litre = 40.000
            full_tank = true

            [[vehicle.fuel]]
            date = 2026-01-01
            odometer_km = 150
            litres = 8.000
            price_per_litre = 40.000
            full_tank = true
        """.trimIndent()

        val tally = repository.importBundle(bundle).vehicles.single()
        assertEquals(2, tally.added.fuel)
        assertEquals(2, repository.loadVehicle("spare").fuel.entries.size)
    }

    @Test
    fun `service dedup key needs part — tyres and an oil change, same day and reading, both import`() {
        val repository = repository()
        val bundle = """
            format = "tritium-export"
            format_version = 1
            exported = 2026-08-19
            source = "android"

            [[vehicle]]
            slug = "spare"
            name = "Spare"

            [[vehicle.service]]
            date = 2026-01-01
            part = "Oil change"
            odometer_km = 500
            amount = 100.00
            vendor = ""

            [[vehicle.service]]
            date = 2026-01-01
            part = "Tyres"
            odometer_km = 500
            amount = 800.00
            vendor = ""
        """.trimIndent()

        val tally = repository.importBundle(bundle).vehicles.single()
        assertEquals(2, tally.added.service)
        assertEquals(2, repository.loadVehicle("spare").service.entries.size)
    }

    @Test
    fun `cost dedup key is date, category and amount — a different amount is a different bill`() {
        val repository = repository()
        val bundle = """
            format = "tritium-export"
            format_version = 1
            exported = 2026-08-19
            source = "android"

            [[vehicle]]
            slug = "spare"
            name = "Spare"

            [[vehicle.costs]]
            date = 2026-01-01
            group = "manual"
            category = "yikama"
            amount = 500.00
            income = false

            [[vehicle.costs]]
            date = 2026-01-01
            group = "manual"
            category = "yikama"
            amount = 600.00
            income = false
        """.trimIndent()

        val tally = repository.importBundle(bundle).vehicles.single()
        assertEquals(2, tally.added.costs)
    }

    @Test
    fun `a vehicle with no records at all is legal and imports with zero counts`() {
        val repository = repository()
        val bundle = """
            format = "tritium-export"
            format_version = 1
            exported = 2026-08-19
            source = "android"

            [[vehicle]]
            slug = "spare"
            name = "Spare"
        """.trimIndent()

        val tally = repository.importBundle(bundle).vehicles.single()
        assertTrue(tally.vehicleCreated)
        assertEquals(Counts(0, 0, 0), tally.added)
        assertEquals("Spare", repository.loadVehicle("spare").vehicle?.vehicle?.name)
    }

    @Test
    fun `refusing a bundle leaves every file on disk exactly as it was`() {
        val repository = repository()
        repository.importBundle(newVehicleBundle)
        val before = paths().fuelToml("sportage").readText()

        assertThrows(BundleError::class.java) {
            repository.importBundle("format = \"not-tritium\"\nformat_version = 1\n")
        }

        assertEquals(before, paths().fuelToml("sportage").readText())
    }

    @Test
    fun `a bundle from a newer format_version is refused, nothing written`() {
        val repository = repository()
        val bundle = """
            format = "tritium-export"
            format_version = 99
            exported = 2026-08-19
            source = "android"

            [[vehicle]]
            slug = "spare"
            name = "Spare"
        """.trimIndent()

        assertThrows(BundleError::class.java) { repository.importBundle(bundle) }
        assertEquals(emptyList<String>(), repository.listVehicleSlugs())
    }

    @Test
    fun `malformed TOML text is refused as unreadable, nothing written`() {
        val repository = repository()
        assertThrows(BundleError::class.java) { repository.importBundle("not valid [[[") }
        assertEquals(emptyList<String>(), repository.listVehicleSlugs())
    }

    @Test
    fun `a corrupt existing vehicle file aborts the whole import — nothing half-applied`() {
        val repository = repository()
        repository.saveVehicleRecord("first", VehicleDocument(1, EMPTY_VEHICLE.copy(name = "First"), emptyMap()))
        File(tmp.root, "vehicles/second").mkdirs()
        File(tmp.root, "vehicles/second/vehicle.toml").writeText("schema_version = 1\nname = \"Second\"\n")
        File(tmp.root, "vehicles/second/fuel.toml").writeText("not valid [[[")

        val bundle = """
            format = "tritium-export"
            format_version = 1
            exported = 2026-08-19
            source = "android"

            [[vehicle]]
            slug = "first"
            name = "First"

            [[vehicle.fuel]]
            date = 2026-01-01
            odometer_km = 100
            litres = 10.000
            price_per_litre = 40.000
            full_tank = true

            [[vehicle]]
            slug = "second"
            name = "Second"
        """.trimIndent()

        assertThrows(CorruptRecordException::class.java) { repository.importBundle(bundle) }

        // "first" sorts earlier in the bundle and would have gained an
        // entry — but the whole import aborts before any write happens.
        assertEquals(emptyList<FuelEntry>(), repository.loadVehicle("first").fuel.entries)
    }

    @Test
    fun `no backup round for a freshly created vehicle, nor for one entirely skip-wins-deduped`() {
        val repository = repository()
        assertFalse(paths().backupsDir.isDirectory)

        repository.importBundle(newVehicleBundle)
        assertFalse(paths().backupsDir.isDirectory)

        repository.importBundle(newVehicleBundle)
        assertFalse(paths().backupsDir.isDirectory)
    }

    @Test
    fun `a backup round is created for an existing vehicle that gains a genuinely new entry`() {
        val repository = repository()
        repository.importBundle(newVehicleBundle)
        assertFalse(paths().backupsDir.isDirectory)

        val secondBundle = """
            format = "tritium-export"
            format_version = 1
            exported = 2026-08-20
            source = "android"

            [[vehicle]]
            slug = "sportage"
            name = "SPORTAGE 1.6 T-GDI"

            [[vehicle.fuel]]
            date = 2026-09-01
            odometer_km = 20000
            litres = 30.000
            price_per_litre = 75.000
            full_tank = true
        """.trimIndent()

        repository.importBundle(secondBundle)
        assertTrue(paths().backupsDir.isDirectory)
    }
}
